### ISSUE TITLE
Avoid data objects serialization

### ISSUE THREAD

**Issue by mkcosta** [checkout/checkout-sdk-ruby#99]
### Environment

* Checkout SDK version: 1.0.3
* Platform and version: N/R
* Operating System and version: N/R

### Description

Ruby does not need data objects to represent `JSON` data, it already has the `Hash` one: i suggest allowing to pass also `Hash` objects to the APIs client, thus avoiding instantiating data objects, serialize them as hashes, spare verbosity and improve performances (creating objects and traverse their instance variables is slow in Ruby). 

## Proposed Solution

```ruby
module CheckoutSdk
  class JsonSerialize
    ...
    def self.to_custom_hash(object)
      return object if object.is_a?(Hash)
      hash = {}
      object.instance_variables.each do |v|
        value = object.instance_variable_get(v)
        value = serialize_by_type(value)
        hash[keys_transformation(v)] = value
      end
      hash
    end
    ...
  end
```

**Comment by Armando-Ibarra** [comment]
Hi @mkcosta.

There are a few reasons why we decided to serialize the response, some of those reasons are the diversity of the response types.
Can you explain how this really affects the performance of calling a few endpoints and serializing the response?

I can imagine a client processing millions of transactions per second, but we don't really see a big performance loss using OpenStruct.

**Comment by mkcosta** [comment]
Hi @Armando-Ibarra,thanks for feedback.

As said: Ruby dosen't need specific data structures (like Java, GO, etc) to represents data objects. It has Array and Hashe classes that serve the purpose, are flexible and can be converted/parsed quickly.

Your implementation not only is slower and verbose (see below), but forces the programmer using your client to know about all of your data structures (pretty convoluted). Also: you didn't use OpenStruct here, but plain classes without any constructor (mostly): this is not only verbose, but force to use accessors (slower) without any of the benefit they provide (data manipulation).

Here you can find a small benchmark i've done by using an Hash and your data structures (also compare the verbosity):
```ruby
# frozen_string_literal: true

require 'benchmark/ips'
require_relative '../../config/environment'

data = {:source=>{:type=>'token', :token=>'tok_ubfj2q76miwundwlk72vxt2i7q', :billing=>{:address=>{:address_line1=>'Viale Borgo Valsugana 93', :city=>'Prato', :state=>'PO', :zip=>'59100', :country=>'IT'}}}, :amount=>5000, :capture=>false, :currency=>'EUR', :payment_type=>'Regular', :reference=>'Order #6362', :customer=>{:email=>'filippo@commercelayer.io', :name=>'Filippo Conforti'}, :shipping=>{:address=>{:address_line1=>'Viale Borgo Valsugana 93', :city=>'Prato', :state=>'PO', :zip=>'59100', :country=>'IT'}}, :'3ds'=>{:enabled=>true}, :success_url=>'https://sandbox.checkout.com/api2/v2/3ds/acs/sid_feixbit6us3utfedjulm6egnsu'}

object = CheckoutSdk::Payments::PaymentRequest.new.tap do |obj|
  obj.amount = data[:amount]
  obj.capture = data[:capture]
  obj.currency = CheckoutSdk::Common::Currency::EUR
  obj.payment_type = data[:payment_type]
  obj.reference = data[:reference]
  obj.success_url = data[:success_url]
  obj.source = CheckoutSdk::Payments::TokenSource.new.tap do |src|
    src.token = data[:source][:token]
    src.billing_address = CheckoutSdk::Common::Address.new.tap do |addr|
      addr.address_line1 = data[:source][:billing][:address][:address_line1]
      addr.city = data[:source][:billing][:address][:city]
      addr.state = data[:source][:billing][:address][:state]
      addr.zip = data[:source][:billing][:address][:zip]
      addr.country = data[:source][:billing][:address][:country]
    end
  end
  obj.customer = CheckoutSdk::Common::CustomerRequest.new.tap do |cus|
    cus.email = data[:customer][:email]
    cus.name = data[:customer][:name]
  end
  obj.shipping = CheckoutSdk::Common::Address.new.tap do |addr|
    addr.address_line1 = data[:source][:billing][:address][:address_line1]
    addr.city = data[:shipping][:address][:city]
    addr.state = data[:shipping][:address][:state]
    addr.zip = data[:shipping][:address][:zip]
    addr.country = data[:shipping][:address][:country]
  end
  obj.three_ds = CheckoutSdk::Payments::ThreeDSRequest.new.tap do |tds|
    tds.enabled = true
  end
end

Benchmark.ips do |x|
  x.report('hash') do
    CheckoutSdk::JsonSerializer.to_custom_hash(data)
  end

  x.report('object') do
    CheckoutSdk::JsonSerializer.to_custom_hash(object)
  end

  x.compare!
end
````

Here's the result on my Macbook Pro 14 10cores:
```
ruby spec/models/checkout_com_client_bench.rb 
Warming up --------------------------------------
                hash     1.465M i/100ms
              object     7.022k i/100ms
Calculating -------------------------------------
                hash     14.617M (± 0.8%) i/s -     73.254M in   5.011990s
              object     70.191k (± 2.5%) i/s -    351.100k in   5.005120s

Comparison:
                hash: 14616772.4 i/s
              object:    70190.6 i/s - 208.24x  (± 0.00) slower
````

As you can see there are two order of magnitudes (not counting the memory consumption). 
You're probably right: in a context of a big application with millions of HTTP requests per day (as ours) the difference will be probably negligible. This will not justify using a verbose, less efficient and non-Rubyesque code.

Also consider we already have an impedance mismatching when composing a payload with our internal data. We definitively don't need your data structure for that, the Hash is a much better solution (we use a refinements btw).

I can understand you'd like to preserve concistency with the client implementation in the other languages, but, as said, Ruby is quite different (Python too).

**Comment by mkcosta** [comment]
Here's the memory profiling too:
```
Calculating -------------------------------------
                hash     0.000  memsize (     0.000  retained)
                         0.000  objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
              object     4.568k memsize (     0.000  retained)
                        64.000  objects (     0.000  retained)
                        42.000  strings (     0.000  retained)

Comparison:
                hash:          0 allocated
              object:       4568 allocated - Infx more
```

**Comment by Armando-Ibarra** [comment]
@mkcosta I see yes, I thought you were referring to response objects because we already had a similar conversation. One of the initial approaches was to provide the request objects, mostly to avoid some unexpected errors on requests, but I understand your point and agree that we should have allowed to instantiate the request either by constructor or by simple hashing and this is one of the things that was escaped during our initial approaches.

Since this may cause breaking changes for some merchants, we are going to change the serializer to accept hashes parameters. Later on, we can either remove the request object or improve the way of how it is generated.

Thanks for your feedback and your patience explaining this.

**Comment by Armando-Ibarra** [comment]
Hey @mkcosta this was released on #108 
If you face any other issue, please let us know.
